Rando: Item Fanfares

Sarge-117 opened the pull request:
In rando, when we find a special quest item somewhere (e.g. a medallion, stone, or song), play that item's unique fanfare instead of the default "get item" fanfare.  Changes tested on items in chests, freestanding items, and as both child and adult.

Closes  HarbourMasters/Shipwright#716

**Edit July 24**
Made some changes to account for getting items from Skulltulas in tokensanity.  Small item fanfares for rupees, heart containers, and normal items will now play when getting them from Skulltulas (instead of being the "get small item" fanfare no matter what).  Getting a token from a chest will also play "get small item" instead of the normal "get item" - matching the sound they would play when picked up from a Skulltula.

Footage: https://cdn.discordapp.com/attachments/988962246585634856/1000468698127155291/tokensanity_items.mp4

Longer fanfares for medallions/stones/songs remain behind the cvar ~~"gRandoFanfareByItemType"~~ "gRandoQuestItemFanfares", which is can currently only be enabled by manually adding it to the json file.  Eventually it can go with other rando-only non-save file cvars when they have a home.
**Edit** It will have a home in PR #915

Sarge-117 committed:
Rando: Minor feature - quest item fanfares

In rando, when we find a special quest item somewhere (e.g. a medallion, stone, or song), play that item's unique fanfare instead of the default "get item" fanfare

Closes  HarbourMasters/Shipwright#716

Sarge-117 committed:
Fix Zelda's Lullaby

Accidentally had prelude in there twice

PurpleHato commented:
I think this should be at best a togglable, these fanfares are way to long and were defined for cutscene purpose and not for gameplay usage with pick-up logic

Sarge-117 commented:
Thanks for the feedback @PurpleHato!  The length did occur to me, so maybe I should elaborate on my reasoning for requesting this feature.  I like it because:

- Provides unique audio feedback for these often-important items
- Distinct fanfare makes it easy to know what you just picked up (for those players who are in a hurry or who are mashing through textboxes, perhaps absent-mindedly or in a race)
- Allows these items to reclaim some of their lost "identity" (since their associated cutscenes are skipped in rando)
- Breaks up the monotony of hearing the same handful of jingles over and over again

In the testing I've done, it works out well in most cases because the fanfares will naturally play themselves out as the player resumes control after picking up the item.  It gets clunky when you have multiple item checks very close together (i.e. Mido's house) and you end up with one fanfare abruptly interrupting another (this is what happens in N64 rando when you shuffle music and end up with longer item fanfares).  The worst offender for length (getting a spiritual stone) also happens to be the rarest, being heard at most three times in a playthrough, which is a convenient coincidence.

Another case is when the fanfare is still playing during a scene transition (which would often be the case when these items are left as dungeon rewards from the blue warp).  In this case the fanfare fades out along with the scene (unless the player has a slow text speed or lingers on the textbox for a moment).  This is similar to how a warp song will fade out prematurely after the player presses "yes" to warp somewhere.

I found overall that this change worked out better than I had planned, with only one scenario sounding awkward (getting another item immediately after one of these).  I like letting these items be "special" and distinct from all the other items in their own way.  But more feedback and discussion is always welcome!

MelonSpeedruns commented:
I do agree with Sarge here, and I do think this change is great for consistency with the vanilla game. I _think_ you could shorten the code by checking if the `itemId >= GI_MEDALLION_FOREST` and `itemId <= GI_MEDALLION_LIGHT` since the medallions are next to each other in the array. Same thing for stones & songs.

PurpleHato commented:
I still don't like the idea to get theses lenghty fanfare over a simple "tadadadaaaa", It's a really bad idea

MelonSpeedruns commented:
I do agree with that. @Sarge-117 Could you add a checkbox in the randomizer options to the likes of `Disable Long Fanfares` for the feature? Make it so the checkbox is disabled by default.

Sarge-117 commented:
A toggleable option for the player sounds great to me.  I'm a bit of a noob so it will take me some time to find all the ui/ux and cvars related stuff, but I'll give it a go.  Thanks @MelonSpeedruns

PurpleHato commented:
> I do agree with that. @Sarge-117 Could you add a checkbox in the randomizer options to the likes of `Disable Long Fanfares` for the feature? Make it so the checkbox is disabled by default.

Since it's not a vanilla behavior, I would more do it the other way around
"Enable lenghty fanfare"

aMannus commented:
Since this is a rando specific change, I thnk it would have to go into the randomizer settings menu. The thing is, so far, anything rando (settings) related has been linked to your save file. While this option sounds like something you'd want to possibly toggle on the fly.

So I'm not entirely sure what the best way to go about it would be, to be honest. Having only some options on a "realtime" toggle while others aren't could get pretty confusing. But it also doesn't feel right to link this to save generation.

I'm not saying it shouldn't be done, I just don't know the best way to implement it would be.

Also, throwing in my own opinion on it, I personally think the option should be off by default.

PurpleHato commented:
> Since this is a rando specific change, I thnk it would have to go into the randomizer settings menu. The thing is, so far, anything rando (settings) related has been linked to your save file. While this option sounds like something you'd want to possibly toggle on the fly.
> 
> So I'm not entirely sure what the best way to go about it would be, to be honest. Having only some options on a "realtime" toggle while others aren't could get pretty confusing. But it also doesn't feel right to link this to save generation.
> 
> I'm not saying it shouldn't be done, I just don't know the best way to implement it would be.

I think the Cvar under the rando options is enough, we will never found a wild spiritual stone, medallion, or song in the vanilla game, but then it will be up to people when they play rando  to remember they had that dfeature enabled on the rando settings, which is why I think the best case it to have it called "Enable item type fanfare" unchecked by default under a certain Rando tab

aMannus commented:
> > Since this is a rando specific change, I thnk it would have to go into the randomizer settings menu. The thing is, so far, anything rando (settings) related has been linked to your save file. While this option sounds like something you'd want to possibly toggle on the fly.
> > So I'm not entirely sure what the best way to go about it would be, to be honest. Having only some options on a "realtime" toggle while others aren't could get pretty confusing. But it also doesn't feel right to link this to save generation.
> > I'm not saying it shouldn't be done, I just don't know the best way to implement it would be.
> 
> I think the Cvar under the rando options is enough, we will never found a wild spiritual stone, medallion, or song in the vanilla game, but then it will be up to people when they play rando to remember they had that dfeature enabled on the rando settings, which is why I think the best case it to have it called "Enable item type fanfare" unchecked by default under a certain Rando tab

I'm not sure how this is related to the vanilla game? I'm just saying I don't like this option being baked into a save or having an option that isn't baked into a save in a menu where every option other than this one is baked into the save.

PurpleHato commented:
> > > Since this is a rando specific change, I thnk it would have to go into the randomizer settings menu. The thing is, so far, anything rando (settings) related has been linked to your save file. While this option sounds like something you'd want to possibly toggle on the fly.
> > > So I'm not entirely sure what the best way to go about it would be, to be honest. Having only some options on a "realtime" toggle while others aren't could get pretty confusing. But it also doesn't feel right to link this to save generation.
> > > I'm not saying it shouldn't be done, I just don't know the best way to implement it would be.
> > 
> > 
> > I think the Cvar under the rando options is enough, we will never found a wild spiritual stone, medallion, or song in the vanilla game, but then it will be up to people when they play rando to remember they had that dfeature enabled on the rando settings, which is why I think the best case it to have it called "Enable item type fanfare" unchecked by default under a certain Rando tab
> 
> I'm not sure how this is related to the vanilla game? I'm just saying I don't like this option being baked into a save or having an option that isn't baked into a save in a menu where every option other than this one is baked into the save.

Best solution would be to make the option and to bake it then into the spoilerlog, I guess, does that sounds reasonable to you ?
Would be easier to discuss if Sarge was on the discord

aMannus commented:
> Best solution would be to make the option and to bake it then into the spoilerlog, I guess, does that sounds reasonable to you ?

That's the exact same thing though, just a different place. It just feels like this is something you might want to toggle on the fly. Ultimately I don't really feel that strongly about this either way, so if everybody else is fine with just putting it in the rando settings and keeping it the same throughout a whole save file, that's fine by me.

Sarge-117 committed:
Add logic to also check for a toggle setting

Added && for the gRandoFanfareByItemType to allow this to be a toggle setting

Sarge-117 committed:
Fix syntax

I think the !=0 is needed

Sarge-117 commented:
These changes to z_player.c should be ready to go for when a suitable spot in the UI is set up and linked to the cvar "gRandoFanfareByItemType"

Sarge-117 committed:
Merge branch 'HarbourMasters:develop' into rando-quest-item-fanfares

Sarge-117 committed:
Merge branch 'HarbourMasters:develop' into rando-quest-item-fanfares

Sarge-117 committed:
Allow Equipment Toggle

Allow player to toggle equipment on/off on the equipment subscreen.  For tunics and boots, this will revert them to Kokiri Tunic/Kokiri Boots.  For shields, it will un-equip the shield entirely.  For swords, only BGS/Giant's Knife is affected, and it will revert to Master Sword.

Sarge-117 committed:
Set up Cvar and check for MS

Set up the cvar/imgui and add a check that we own the Master Sword (in case we're in the Ganon fight)

Sarge-117 committed:
Merge pull request #2 from Sarge-117/rando-quest-item-fanfares

Rando quest item fanfares

Sarge-117 committed:
Merge pull request #3 from Sarge-117/Allow-Equipment-Toggle

Allow equipment toggle

Sarge-117 committed:
Giant's Knife Behaviour Fixes

Fixes a case where Giant's Knife (specifically, breaking and re-buying it) can behave unexpectedly if you don't have a Kokiri Sword in your inventory.  Also fixes the broken icon not showing up in inventory after you break it.

Sarge-117 committed:
less whitespace

Sarge-117 committed:
Merge pull request #4 from Sarge-117/fix-broken-knife-inventory-icon

Fix broken knife inventory icon

Sarge-117 committed:
Merge branch 'HarbourMasters:develop' into rando-quest-item-fanfares

Sarge-117 committed:
Merge branch 'HarbourMasters:develop' into test

Sarge-117 committed:
Shorten code

Shortened the code

Sarge-117 committed:
Merge branch 'test' into rando-quest-item-fanfares

Sarge-117 committed:
Undo accidental commit

I'm a dummy and accidentally committed changes from other branches I was working on.

Sarge-117 commented:
Accidentally committed changes from other branches I was working on - oops lol.  I reverted those and also shortened the code per @MelonSpeedruns suggestion (thanks!).  Tested again and still functions as expected.

Sarge-117 committed:
Merge branch 'HarbourMasters:develop-zhora' into rando-quest-item-fanfares

Sarge-117 committed:
Merge branch 'HarbourMasters:develop-zhora' into rando-quest-item-fanfares

Sarge-117 committed:
Merge pull request #6 from Sarge-117/rando-next

Rando next

Sarge-117 committed:
Account for tokensanity

Added changes to restore small item fanfares to rupees, heart containers, and items when picked up from skulltulas in tokensanity.

Sarge-117 commented:
Changed target branch to rando-next and added changes to restore small item fanfares in tokensanity (for rupees, hearts, and normal items).  The long quest item fanfares are still behind the Cvar.  See update to the main PR description above for footage.

Marking this as ready for review again since the smaller item fanfares are good to go.

Sarge-117 committed:
Account for boomerang/hookshot tokens

Fixed the case where you get the token via the boomerang or hookshot.  Done by separating out the code into its own function and then calling it in the two appropriate spots.

Sarge-117 commented:
Thank you to @aMannus for his PR #915 which adds a spot for the cvar!

PurpleHato commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12699,6 +12699,30 @@ s32 func_8084DFF4(GlobalContext* globalCtx, Player* this) {
+                        temp1 = NA_BGM_OCA_FAIRY_GET | 0x900;
+                    }
+                }
+            }// ***************************************************************************************
```
```suggestion
            }
```

I would remove these kinds of unnecessary comment since we don't tend to do that ^-^
Great PR anyway!

Sarge-117 committed:
Remove comment border

Co-authored-by: PurpleHato <linkvssangoku.jr@gmail.com>

Sarge-117 commented:
Done @PurpleHato :)

aMannus commented:
The only thing I've kinda been thinking about, would it make sense to put the skull tokens part behind the CVar after all? It might be kind of confusing that skulls are already changed without enabling the option, while the rest of the checks (that might feel similar to the player) will have it locked behind the CVar.

Sarge-117 commented:
To me it makes sense to leave the small ones as is, since it brings them in line with the other checks.  E.G. if you get a blue rupee it will play the "got a rupee" sound, whether you get it from a chest or from a skulltula.  Same for heart containers (or the 4th heart piece) and for regular items.

So, if this was disabled, then you would hear "got a rupee" if you get it from a chest, but you would hear "get small item" if you get it from a skulltula.  Obviously not the end of the world, but it feels inconsistent to me.

The long ones for medallions/stones/songs are the special case because the only precedent they have is from cutscenes during the vanilla game.

aMannus commented:
> To me it makes sense to leave the small ones as is, since it brings them in line with the other checks. E.G. if you get a blue rupee it will play the "got a rupee" sound, whether you get it from a chest or from a skulltula. Same for heart containers (or the 4th heart piece) and for regular items.
> 
> So, if this was disabled, then you would hear "got a rupee" if you get it from a chest, but you would hear "get small item" if you get it from a skulltula. Obviously not the end of the world, but it feels inconsistent to me.
> 
> The long ones for medallions/stones/songs are the special case because the only precedent they have is from cutscenes during the vanilla game.



Oohh, I'm sorry then. I honestly didn't even know that already happened for items from chests etc. And I see now that the longer fanfares are still locked behind the CVar even when picked up from skulls. That does mean I should probably update the name and tooltip for the CVar then though, because it doesn't really make sense what I've written for it then.

Sarge-117 committed:
Change cvar name for accuracy

Changed Cvar from "gRandoFanfareByItemType" to "gRandoQuestItemFanfares" for better accuracy

Sarge-117 committed:
Fix special case with the "WINNER" heart

The "WINNER" heart is its own unique item rather than just another piece of heart, so it needed its own check to make it consistent with other pieces of heart

Sarge-117 commented:
Did some additional testing today and it looks like these two functions are the only ones that get called for items obtained from chests, from the ground, from skulltulas, with the hookshot/boomerang, from underwater, from cutscenes, and from shops and scrubs.  So this should be compatible with future shopsanity/scrubsanity settings.

Also fixed a special case with the "WINNER" heart piece, which is its own unique item, so it was getting missed by the other conditionals.

aMannus commented:
Would it make sense to use `Audio_PlayFanfare_Rando();` inside of `z_player.c` as well? Seeing as they do nearly identical things. Would make it more consistent for the future if we ever need to call the audio for anything we add in the future (new sanities etc).

Sarge-117 commented:
@aMannus Only a portion of the code is relevant to rando in z_player.c (the part that handles the tokens in tokensanity and the part that handles the optional longer fanfares).  The rest is vanilla code for handling which fanfare to play when Link gets an item. 

In z_en_si.c, basically the entire thing (the vanilla part from z_player.c plus the additional stuff that I wrote) was copied over to create the `Audio_PlayFanfare_Rando()` function, since normally z_en_si.c is only calling for the "get small item" fanfare, and never had any logic of its own for changing which one to play.

Basically in z_player.c, I had to make changes to account for a few special cases that only happen in rando.  But in z_en_si.c, I had to make changes to account for every possible item in the game now possibly coming from a skull, haha.  So for now I've left it as is, because reducing them both to one function call would involve making changes to vanilla code that gets called outside of rando.

**Edit:**  Sorry @aMannus I was being a dummy and I realized I could just call the function at the end and skip that chunk of z_player.c lol.  I am testing it out now!

Sarge-117 committed:
Shorten code by calling new function

Reduced duplicate code by calling the new Audio_PlayFanfare_Rando function from z_player.c

Sarge-117 commented:
Updated to shorten the code and call the function from z_player.c (only for n64ddFlag) as well - thank you @aMannus for the suggestion!  It works well!

Sarge-117 committed:
Fix declaration

Forgot to include the argument in the declaration at the top.  It worked fine even without it, but figured this is safer

aMannus commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -12699,7 +12704,13 @@ s32 func_8084DFF4(GlobalContext* globalCtx, Player* this) {
             else {
                 temp1 = temp2 = (this->getItemId == GI_HEART_PIECE) ? NA_BGM_SMALL_ITEM_GET : NA_BGM_ITEM_GET | 0x900;
             }
-            Audio_PlayFanfare(temp1);
+          
+            RandoFanfareSkip:
+            if (gSaveContext.n64ddFlag) {
+                Audio_PlayFanfare_Rando(this->getItemId);
+            } else {
+                Audio_PlayFanfare(temp1);
+            }
```
Nitpick. You could leave the normal `Audio_PlayFanfare(temp1);` above `RandoFanFareSkip:` and only keep the if statement for the rando fanfare.

briaguya0 replied:
but then it would play both the normal and the rando fanfare

briaguya0 replied:
oh wait, gotos? oof

aMannus replied:
There's this higher up in the code:

```
// In rando the fanfares are handled in z_en_si.c, so we can skip down a bit
        if (gSaveContext.n64ddFlag) {
            goto RandoFanfareSkip;
        }
```

I was under the impression that skipped the code until that was met.

briaguya0 replied:
this would be better if it was just:
```
if rando and rando fanfares {
 play rando fanfare
} else if {
 rupee/heart stuff
} else {
 other item stuff
}
```

Sarge-117 replied:
Sounds good!  I will work on this and test again tonight.

briaguya0 replied:
> There's this higher up in the code

yeah, noticed that, hence my "gotos oof" comment

aMannus replied:
Yeah we posted it at the same time. The conditionals are ready happening in the new rando function though. Shouldn't wrapping the z_player.c stuff in a single if/else be enough instead of a goto then?

The reason why I suggested a generic rando function before is so it can catch things like ice traps or future custom items as well.

Sarge-117 committed:
Code cleanup

Some optimizations per suggestions

Sarge-117 commented:
Shortened the code - thank you @briaguya-ai @aMannus for the feedback!  I'm tempted to say this might actually be ready now!

briaguya0 commented on soh/src/overlays/actors/ovl_En_Si/z_en_si.c:
```diff
@@ -149,11 +155,58 @@ void func_80AFB89C(EnSi* this, GlobalContext* globalCtx) {
 
+// Function to play "get-item" fanfares according to the type of item obtained (used in rando)
+// Longer fanfares for medallions/stones/songs are behind the Cvar
+void Audio_PlayFanfare_Rando(ItemID getItemId) {
```
seems really odd to have this living in the skull token actor, not sure the best spot for it but i feel like even `OTRGlobals` would be better

Sarge-117 replied:
The normal `Audio_PlayFanfare` and `Audio_PlaySoundGeneral` were already defined and used in `code_800EC960.c` so putting it there worked out without any trouble!

briaguya0 reviewed (changes requested):
`Audio_PlayFanfare_Rando` needs to move out of the skull token actor, in a previous comment i mentioned even `OTRGlobals` being a better place for it, but since the vanilla fanfare function is in `code_800EC960` putting it there might work too

Sarge-117 committed:
Move function to code_800EC960.c

Moved Audio_PlayFanfare_Rando to a better home

briaguya0 commented on soh/src/code/code_800EC960.c:
```diff
@@ -3887,6 +3888,49 @@ void Audio_ResetSfxChannelState(void) {
+void Audio_PlayFanfare_Rando(ItemID getItemId) {
+    s32 temp1;
+
+    if (((getItemId >= GI_RUPEE_GREEN) && (getItemId <= GI_RUPEE_RED)) ||
+        ((getItemId >= GI_RUPEE_PURPLE) && (getItemId <= GI_RUPEE_GOLD)) ||
+        ((getItemId >= GI_RUPEE_GREEN_LOSE) && (getItemId <= GI_RUPEE_PURPLE_LOSE)) || (getItemId == GI_HEART)) {
+        Audio_PlaySoundGeneral(NA_SE_SY_GET_BOXITEM, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+    } else {
+        if ((getItemId == GI_HEART_CONTAINER_2) || (getItemId == GI_HEART_CONTAINER) ||
+            ((getItemId == GI_HEART_PIECE) && ((gSaveContext.inventory.questItems & 0xF0000000) == 0x40000000))) {
+            temp1 = NA_BGM_HEART_GET | 0x900;
+        } else {
+            temp1 = (getItemId == GI_HEART_PIECE) ? NA_BGM_SMALL_ITEM_GET : NA_BGM_ITEM_GET | 0x900;
+        }
```
i know this part is all copypasta from `func_8084DFF4` in `z_player.c`

it would be nice to consolidate the logic at some point but i don't think that needs to happen in this PR

aMannus commented:
LGTM!
